File: src/ollie/session.py
from __future__ import annotations

from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from ollie.client import Client


class SessionContext:
    """Optional grouping context; sets session_id on nested workflows."""

    def __init__(self, client: Client, session_id: str) -> None:
        self._client = client
        self.session_id = str(session_id).strip()
        self._previous: str | None = None

    def __enter__(self) -> SessionContext:
        self._previous = self._client._session_id
        self._client._session_id = self.session_id
        return self

    def __exit__(self, *args: object) -> None:
        self._client._session_id = self._previous

File: packages/ts/src/session.ts
import type { Client } from "./client.js";

export class SessionContext {
  private readonly client: Client;
  readonly sessionId: string;
  private previous: string | null = null;

  constructor(client: Client, sessionId: string) {
    this.client = client;
    this.sessionId = sessionId.trim();
  }

  enter(): this {
    this.previous = this.client.sessionId;
    this.client.sessionId = this.sessionId;
    return this;
  }

  exit(): void {
    this.client.sessionId = this.previous;
  }

  [Symbol.dispose](): void {
    this.exit();
  }
}
